File: Algorithm/Gemoetry/ConvexHull(GrahamScan).cpp
#include <algorithm>
#include <vector>
#include <iostream>
 
using namespace std;
typedef long long ll; 

typedef struct __point{
    ll x, y;
} Point;

vector<Point> P;
Point ref_p;

int CCW(const Point& A, const Point& B, const Point& C)
{
    ll cross_product = (B.x - A.x)*(C.y - A.y) - (C.x - A.x)*(B.y - A.y);
    if (!cross_product) return 0;
    return cross_product > 0 ? 1 : -1;
}

bool order_asc_yx(const Point& A, const Point& B)
{
    if (A.y != B.y) return A.y < B.y;        // y가 asc이면 swap하지 않음(=true)
    else return A.x < B.x;                   // y가 동일할 때 x가 asc이면 swap하지 않음(=true)
}

bool order_ccw(const Point& A, const Point& B)
{
    int dir = CCW(ref_p, A, B);
    if (dir == 0) return order_asc_yx(A, B); // 직선상에 위치하면 asc y, asc x를 기준으로 정렬
    else return dir == 1;                    // CCW이면 swap을 하지 않음(=true), CW이면 swap (=false)
}

int GrahamScan(int N)
{
    sort(P.begin(), P.end(), order_asc_yx);
    ref_p = P[0];
    sort(P.begin() + 1, P.end(), order_ccw);
    
    vector<Point> stack = { ref_p, P[1] };
    for (int i = 2; i < N; ++i) {
        int top_idx = stack.size() -1;
        while (stack.size() > 1 && CCW(stack[top_idx -1], stack[top_idx], P[i]) <= 0) {
            stack.pop_back();
            --top_idx;
        }
        stack.push_back(P[i]);
    }
    return stack.size();
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    int N;
    cin >> N;
    for (int i = 0, x, y; i < N; ++i) {
        cin >> x >> y;
        P.push_back({x, y});
    }
    cout << GrahamScan(P.size());
}

File: Algorithm/Gemoetry/ConvexHull(GrahamScan).py
from functools import cmp_to_key
from sys import stdin
input = stdin.readline

class Point :
    def __init__(self, x, y) :
        self.x = x
        self.y = y

def CCW(std, A, B) :
    cross_product = (A.x - std.x)*(B.y - std.y) - (B.x - std.x)*(A.y - std.y);
    if cross_product == 0 : return 0
    return [-1, 1][cross_product > 0]
    
class ConvexHull :
    def __init__ (self) :
        self.ref_point = None
    
    # order asc y asc x
    def orderASC_YX(self, P1, P2) : 
        if P1.y != P2.y : return [-1, 1][P1.y > P2.y]
        return [-1, 1][P1.x > P2.x]
    
    def orderCCW(self, P1, P2) :
        DIR = CCW(self.ref_point, P1, P2)
        if DIR == 0 : return self.orderASC_YX(P1, P2)
        return [1, -1][DIR == 1]    # P1 !== P2
        
    def GrahamScan(self, P, N) :
        P.sort(key = lambda p : (p.y, p.x))
        self.ref_point = P[0]
        del P[0]        # remove reference point 
        P.sort(key = cmp_to_key(self.orderCCW))
        stack = [self.ref_point, P[0]]
        del P[0]        # remove point on a line
        for p in P :
            while len(stack) > 1 and CCW(stack[-2], stack[-1], p) <= 0 :    # CCW(stack_next_top, stack_top, target)
                stack.pop()
            stack.append(p)
        return len(stack)

if __name__ == "__main__" :
    P = [Point(*map(int, input().split())) for _ in range(int(input()))]
    print(ConvexHull().GrahamScan(P, len(P)))
